### PR TITLE
Implements #3474. Add triggered state when a rule is triggered.

### PR DESCRIPTION
**Pull request by danchom**
The running state appears only when actions are executed.

Signed-off-by: Yordan Mihaylov <j.mihailov@prosyst.bg>

### CODE REVIEW

**Comment by danchom** [comment]
@kaikreuzer could you check this PR please

File: bundles/automation/org.eclipse.smarthome.automation.module.timer.test/src/test/groovy/org/eclipse/smarthome/automation/module/timer/internal/RuntimeRuleTest.groovy [hunk 1]
```diff
@@ -133,7 +133,8 @@ class RuntimeRuleTest extends OSGiTest{
          * Check if auto update is working and
          * ensure that the lamp item state if OFF after this check
          */
-        logger.info("Check auto update");
+        logger.info("Check auto update")
```
Comment (sjka) [inline bundles/automation/org.eclipse.smarthome.automation.module.timer.test/src/test/groovy/org/eclipse/smarthome/automation/module/timer/internal/RuntimeRuleTest.groovy]: this line break slipped in

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java [hunk 2]
```diff
@@ -69,6 +74,7 @@ public RuleStatusInfo(RuleStatus status, RuleStatusDetail statusDetail, String d
         this.status = status;
         this.statusDetail = statusDetail;
         this.description = description;
+        this.properties = properties;
```
Comment (sjka) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java]: how do these properties relate to the triggered state?

Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java]: The RuleStatusInfo of triggered state should contain output values of the trigger which was activated the rule. These values can be used by a CEP trigger based on the existing trigger type, also  they can be used for logging or remote tracking of automation engine.

**Comment by danchom** [comment]
Test cases were added.

File: bundles/automation/org.eclipse.smarthome.automation.module.core.test/src/test/java/org/eclipse/smarthome/automation/module/core/internal/TriggeredRunningStateTest.java [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+                        break;
+                    case IDLE:
+                        ruleStatusInfoEventFound = triggeredEventFound;
+                        break;
+                }
+            }
+        } while (!ruleStatusInfoEventFound && wait > 0);
+        Assert.assertTrue(ruleStatusInfoEventFound);
+        Assert.assertNotNull(myLampItemAction);
+        Assert.assertEquals(OnOffType.OFF, myLampItemAction.getState());
```
Comment (triller-telekom) [inline bundles/automation/org.eclipse.smarthome.automation.module.core.test/src/test/java/org/eclipse/smarthome/automation/module/core/internal/TriggeredRunningStateTest.java]: Shouldn't the action lamp be on once the rule has fired?

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatus.java [hunk 4]
```diff
@@ -18,7 +18,8 @@
     UNINITIALIZED(1),
     INITIALIZING(2),
     IDLE(3),
-    RUNNING(4);
+    RUNNING(4),
+    TRIGGERED(5);
```
Comment (triller-telekom) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatus.java]: How about putting "TRIGGERED" before "RUNNING"? Because that is also how the lifecycle of a rule will be

File: bundles/automation/org.eclipse.smarthome.automation.module.core.test/src/test/java/org/eclipse/smarthome/automation/module/core/internal/TriggeredRunningStateTest.java (hunk 3 above)
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.module.core.test/src/test/java/org/eclipse/smarthome/automation/module/core/internal/TriggeredRunningStateTest.java]: Actually no, because the condition is not satisfied (myLampCondition is OFF) and the rule is not executed.  Therefor the action myLampItemAction is not executed and the state is not changed.

**Comment by danchom** [comment]
The PR is updated, it has optimization in rule engine synchronizations also the factory of RuleStatusInfo is added.

**Comment by danchom** [comment]
ping

File: bundles/automation/org.eclipse.smarthome.automation.module.core.test/src/test/java/org/eclipse/smarthome/automation/module/core/internal/TriggeredRunningStateTest.java (hunk 3 above)
Comment (triller-telekom) [inline bundles/automation/org.eclipse.smarthome.automation.module.core.test/src/test/java/org/eclipse/smarthome/automation/module/core/internal/TriggeredRunningStateTest.java]: Ah I see, sorry I missed that condition...

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java [hunk 5]
```diff
@@ -45,20 +75,44 @@ public RuleStatusInfo(RuleStatus status) throws IllegalArgumentException {
...
+    public static RuleStatusInfo create(RuleStatus status, RuleStatusDetail statusDetail) throws IllegalArgumentException {
+        if(statusDetail == null || statusDetail == statusDetail.NONE) {
+            return create(status);
+        }else {
+            return create(status, statusDetail, null);
+        }    
+    }
+
+    public static RuleStatusInfo create(RuleStatus status, RuleStatusDetail statusDetail, String description) {
+        if(status == null) {
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java]: @triller-telekom Instead of having all those IAEs, wouldn't that be a perfect case to make use of the new Nonnull annotations?

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 6]
```diff
@@ -116,40 +120,35 @@
...
      * not. The relation is rule's id to {@link Rule} object.
      */
-    private Map<String, RuntimeRule> rules;
+    private Map<String, RuntimeRule> rules = new ConcurrentHashMap<String, RuntimeRule>();
 
     /**
      * {@link Map} system module type to corresponding module handler factories.
      */
-    private Map<String, ModuleHandlerFactory> moduleHandlerFactories;
+    private Map<String, ModuleHandlerFactory> moduleHandlerFactories = new ConcurrentHashMap<>(31);
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: Any specific reason for 31?

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java (hunk 2 above)
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java]: I will need a deeper look into this, but I am not convinced that such properties should be introduced here (as it brings it out of sync with the Thing status concept and also most likely breaks all our evening mechanism...).

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 7]
```diff
@@ -1,4 +1,8 @@
 /**
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: There are MANY changes in this class - is this really all related to this feature or does it include some further refactorings? In the latter case, I would prefer to have it split into separate PRs, which would make reviews much easier.

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuntimeRule.java [hunk 8]
```diff
@@ -28,6 +33,7 @@
 public class RuntimeRule extends Rule {
 
     private Map<String, Module> moduleMap;
+    private RuleStatusInfo statusInfo;
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuntimeRule.java]: This looks fishy. Afair, we so far kept the status of the rules in the rule engine, having it mixed up with the rule meta-data might not be the best idea...

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java (hunk 7 above)
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: I've optimize synchronizations. Also I have remove status map from rule engine and put it into runtime rule. I don't see why we have to hold a rules' status into separate map and do unduly synchronization between runtime rules and status map. Each  runtime rule must have a status and I think it must be a field of it.

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java (hunk 6 above)
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: No, I'll remove it.

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuntimeRule.java (hunk 8 above)
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuntimeRule.java]: Actually each runtime rule must always have a status. I think it should be part of it. In these case there is no need to synchronize maps of runtime rules and their statuses.

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java [hunk 9]
```diff
@@ -19,6 +26,18 @@
     private RuleStatus status;
     private RuleStatusDetail statusDetail;
     private String description;
+    private Map<String, ?> properties;
+
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java]: I'll

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatus.java [hunk 10]
```diff
@@ -1,4 +1,8 @@
 /**
+ * Copyright (c) 2017 by Bosch Software Innovations GmbH. All rights reserved,
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatus.java]: Well, yes, it is against the agreed rules for the license headers in this project (or rather this component).
If I may repeat them: We use
```
 * Copyright (c) {DATE} {INITIAL COPYRIGHT OWNER} and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
```
where `{INITIAL COPYRIGHT OWNER}` is the party that created the file in the first place (in this case ProSyst) and "and others" includes every one else that has done subsequent changes to it (which imho includes successor companies such as Bosch SI).

We definitely must not have two different license headers on a file.
In any case, please remember that we plan to move to a completely [new style of license headers](https://bugs.eclipse.org/bugs/show_bug.cgi?id=514948) in future.

**Comment by danchom** [comment]
We have a response form our legacy department and it says to change the Prosyst name with the Bosch SI one in the license header when the initial creator is Prosyst and Bosch have modified file after that. We also have to add the year when changing license header.
For example:

/**
 * Copyright (c) 2015 by ProSyst Software GmbH and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

should become to:
/**
 * Copyright (c) 2015, 2017 by Bosch Software Innovations and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

@kaikreuzer is this solution appropriate for you?

**Comment by kaikreuzer** [comment]
Yes, thanks, that's absolutely fine for me.

**Comment by danchom** [comment]
@kaikreuzer Could you please check this PR again.

**Comment by danchom** [comment]
ping

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java (hunk 2 above)
Comment (kaikreuzer) [reply bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java]: No, imho that is not "natural" at all, it is rather the "easy" way... I really think that this is no information that is in any way related to the "RuleStatus". It is some information that you want to send out as an event for specific use cases. Couldn't the eventing be done by a special type of CEPTrigger? Its handler implementation could use the EventAdmin and post there whatever it likes to - without having any negative impact on the general infrastructure.

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuntimeRule.java (hunk 8 above)
Comment (kaikreuzer) [reply bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuntimeRule.java]: Ok, if we consider the RuntimeRules an internal data structure for the rule engine, it should be fine.

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java (hunk 2 above)
Comment (kaikreuzer) [reply bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java]: @adimova Could you follow-up on this? Is it ok for you to move the specific information to the trigger implementation and revert the addition of properties to the RuleStatusInfo from this PR?

Comment (adimova) [reply bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/RuleStatusInfo.java]: Yes, that was my intention, to prepare new PR with "Add triggered state when a rule is triggered", 'cause this PR has merge conflicts, also we still should discuss CEP concept to get clean. So this PR can be closed.

**Comment by kaikreuzer** [comment]
I guess the issue is that this PR had been created by @danchom and @adimova cannot easily take it over. So ok for me to close and have a new one created by @adimova.
